providers/summary/scancode-new.js

```diff
@@ -0,0 +1,181 @@
+    const [firstPackage] = packages
+    if (!firstPackage) return null
+
+    const licenseExpression = firstPackage.declared_license_expression_spdx
+
+    return licenseExpression?.includes('NOASSERTION') ? null : licenseExpression
+  }
+
+  _readExtractedLicenseStatementFromPackage({ content }) {
+    const declared_license = get(content, 'packages[0].extracted_license_statement')
+    return SPDX.normalize(declared_license)
+  }
+
+  // find and return the files that should be considered for as a license determinator for this summarization
+  _getRootFiles(coordinates, files, packages) {
+    const roots = getLicenseLocations(coordinates, packages) || []
+    roots.push('') // for no prefix
+    let rootFiles = this._findRootFiles(files, roots)
+    //Some components (e.g. composer/packgist) are packaged under one directory
+    if (rootFiles.length === 1 && rootFiles[0].type === 'directory') {
+      rootFiles = this._findRootFiles(files, [`${rootFiles[0].path}/`])
+    }
+    return rootFiles
+  }
+
+  _findRootFiles(files, roots) {
+    return files.filter(file => {
+      for (let root of roots) {
+        if (file.path.startsWith(root) && file.path.slice(root.length).indexOf('/') === -1) return true
+      }
+    })
+  }
+
+  _getDetectedLicensesFromFiles(harvestedData, coordinates) {
+    const rootFiles = this._getRootFiles(coordinates, harvestedData.content.files, harvestedData.content.packages)
+    return this._getFileLicensesFromDetectedLicenseExpressions(rootFiles)
+  }
+
+  _getFileLicensesFromDetectedLicenseExpressions(files) {
+    const fullLicenses = files
+      .filter(file => file.percentage_of_license_text >= 90 && file.detected_license_expression_spdx)
+      .map(file => file.detected_license_expression_spdx)
+    return joinExpressions(fullLicenses)
+  }
+
+  _getClosestLicenseMatchByFileName(files, coordinates, licenseClarityScoreThreshold = 90) {
+    const fullLicenses = files
+      .filter(file => isLicenseFile(file.path, coordinates) && file.license_detections)
+      .reduce((licenses, file) => {
+        file.license_detections.forEach(licenseDetection => {
+          if (licenseDetection.license_expression_spdx) {
+            licenses.add(licenseDetection.license_expression_spdx)
+            return
+          }
+          licenseDetection.matches.forEach(match => {
+            if (match.score >= licenseClarityScoreThreshold) {
+              licenses.add(match.spdx_license_expression)
+            }
+          })
+        })
+        return licenses
+      }, new Set())
+    return joinExpressions(fullLicenses)
+  }
+
+  _summarizeFileInfo(files, coordinates) {
+    return files
+      .map(file => {
+        if (file.type !== 'file') return null
+
+        const result = { path: file.path }
+
+        const licenseExpression =
+          file.detected_license_expression_spdx || this._getClosestLicenseMatchByFileName([file], coordinates, 80)
+        setIfValue(result, 'license', licenseExpression)
```

Comment:
hm... licenseExpression for a file can be set even if the file is not a license file for the package (e.g. discovered license).  `isLicenseFile` in `_getClosestLicenseMatchByFileName` may not be necessary here on the file level.  so might be better to refactor the `isLicenseFile `check outside this function.  `isLicenseFile `check is indeed necessary when setting file nature  on line 159.

Reply:
Thank you @qtomlinson!

The old summarizer has this logic:

```javascript
const fileLicense = asserted || file.licenses || []
let licenses = new Set(fileLicense.map(x => x.license).filter(x => x))
if (!licenses.size)
  licenses = new Set(fileLicense.filter(x => x.score >= 80).map(x => this._createExpressionFromLicense(x)))
const licenseExpression = joinExpressions(licenses)
setIfValue(result, 'license', licenseExpression)
```

Using `file.detected_license_expression_spdx` here to me is equivalent to using `joinExpressions(file.licenses)` which is one option that can happen in the old code, would you agree? I'm open to changing this but just wanted to call out that it would be a deviation from the old algorithm.

Can you explain why you think `isLicenseFile` might not be necessary?

Reply:
Thank you @lumaxis for the explanation. Using `file.detected_license_expression_spdx` looks good to me.  `this._getClosestLicenseMatchByFileName([file], coordinates, 80)` branch includes `isLicenseFile` check.  This checks whether the file is a license file for the package, which is not present in the logic of old summarizer. May want to move this `isLicenseFile` check outside `_getClosestLicenseMatchByFileName`.

Reply:
@qtomlinson Yea, good callout 🤔 I've split the previous function up into two, to be able to correctly replicate the old logic correctly. Do you perhaps know of a good test case for this? I'd like to add a unit test where the previous logic would have failed but haven't been able to find one, yet.

Reply:
@lumaxis Thanks for making the change.  I have tried half a dozen test cases where there are discovered license, but was not able to invoke _getLicenseExpressionFromFileLicenseDetections.